### PR TITLE
feature: comtrade goods and services pipelines

### PR DESCRIPTION
**Pull request by michalc**
### Checklist

* [x] Have tests been added to cover any changes?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: tests/unit/operators/test_comtrade.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        },
+    )
+
+    services_csv_2019 = io.BytesIO()
+    with zipfile.ZipFile(services_csv_2019, "w") as zf:
+        zf.writestr(
+            "services_data.csv",
+            (
+                "Classification,Year,Period,Period Desc.,Aggregate Level,Is Leaf Code,Trade Flow Code,Trade Flow,Reporter Code,Reporter,Reporter ISO,Partner Code,Partner,Partner ISO,Commodity Code,Commodity,Trade Value (US$),Flag\n"
+                "S3,2008,2008,2008,1,0,1,Import,4,Afghanistan,AFG,0,World,WLD,0,Food and live animals,309541080,0\n"
```
Comment (abbas123456) [inline tests/unit/operators/test_comtrade.py:128]: I would probably change the classification here to EB to make it look like an actual services csv file

File: dataflow/dags/comtrade_pipelines.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            ('Partner', sa.Column('partner', sa.String, nullable=False)),
+            ('Partner ISO', sa.Column('partner_iso', sa.String)),
+            ('Commodity Code', sa.Column('commodity_code', sa.String, nullable=False)),
+            ('Commodity', sa.Column('commodity', sa.String, nullable=False)),
+            ('Trade Value (US$)', sa.Column('trade_value_usd', sa.Numeric)),
+        ],
+        transforms=[
+            lambda record, table_config, contexts: {
+                **record,
+                'Is Leaf Code': bool(int(record['Is Leaf Code'])),
```
Comment (samuelhwilliams) [inline dataflow/dags/comtrade_pipelines.py:142]: I tend to think it's nice not to overwrite data in these transforms. Maybe you can use a new key like `is_leaf_code_bool`?

File: dataflow/dags/comtrade_pipelines.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+from dataflow.dags import _PipelineDAG
+from dataflow.operators.comtrade import (
+    fetch_comtrade_goods_data,
+    fetch_comtrade_services_data,
+)
+from dataflow.utils import TableConfig
+
+
+class ComtradeGoodsPipeline(_PipelineDAG):
+    allow_null_columns = True
```
Comment (samuelhwilliams) [inline dataflow/dags/comtrade_pipelines.py:15]: What columns are we expecting this on? It shouldn't normally be enabled as it gets rid of some of our (already limited) data quality checks.

File: dataflow/dags/comtrade_pipelines.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class ComtradeServicesPipeline(_PipelineDAG):
+    allow_null_columns = True
+
+    def get_fetch_operator(self) -> PythonOperator:
+        return PythonOperator(
+            task_id="fetch-comtrade-services-data",
+            python_callable=fetch_comtrade_services_data,
+            queue='high-memory-usage',
```
Comment (samuelhwilliams) [inline dataflow/dags/comtrade_pipelines.py:90]: Do these both need the high-memory-usage worker? It would be good to just use the default if it is not strictly required.

File: dataflow/operators/comtrade.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        'Flag',
+    ]
+    _fetch(s3, 'S', expected_keys)
+
+
+def _fetch(s3, trade_type, expected_keys):
+    with _download('https://comtrade.un.org/Data/cache/years.json') as response:
+        years_objs = json.loads(response.content)
+    if years_objs['more']:
+        raise Exception('Unable to fetch all years')
```
Comment (samuelhwilliams) [inline dataflow/operators/comtrade.py:76]: Do we never expect this key to appear? This could probably use either a comment explaining why it's a hard error or just handling the edge(?) case?

File: dataflow/dags/comtrade_pipelines.py (hunk 3 above)
Comment (abbas123456) [reply dataflow/dags/comtrade_pipelines.py:15]: So there were a few null only columns when I was testing locally with a small dataset, i.e not all years or reporters. However we will need to test the entire fetch on dev and if there are still null only columns then they can be removed.

File: dataflow/dags/comtrade_pipelines.py (hunk 4 above)
Comment (abbas123456) [reply dataflow/dags/comtrade_pipelines.py:90]: No the services one doesn't, I think this was a bad copy paste. The goods api zip files are around 500mb and they need to be downloaded to memory before they can be processed. While testing on dev with the normal worker this was fine as the memory gets freed up when it finishes processing each year but it still ran out of memory after around 5 hours. This was possibly due to another pipeline running at the same time so I though it would be best to stick it on the high-memory-usage worker to be on the safe side.

File: dataflow/operators/comtrade.py (hunk 5 above)
Comment (abbas123456) [reply dataflow/operators/comtrade.py:76]: I think this can be removed. I suppose the `more` key will be set to true when there is another page of years but I can't imagine that happening any time soon.

File: dataflow/dags/comtrade_pipelines.py (hunk 3 above)
Comment (samuelhwilliams) [reply dataflow/dags/comtrade_pipelines.py:15]: Ok, so probably sounds like this should be removed? If we merge it to dev with it still there, there's a higher chance we miss it and push it to prod. If we remove it and merge, dev will fail and we will be more likely to catch it?

File: dataflow/operators/comtrade.py (hunk 5 above)
Comment (samuelhwilliams) [reply dataflow/operators/comtrade.py:76]: Ah ok. Maybe it is better to leave it to hard error in the future rather than silently ignore additional years...

File: dataflow/dags/comtrade_pipelines.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            lambda record, table_config, contexts: {
+                **record,
+                'is_leaf_code_bool': bool(int(record['Is Leaf Code'])),
+            }
+        ],
+    )
+
+
+class ComtradeServicesPipeline(_PipelineDAG):
+    allow_null_columns = True
```
Comment (samuelhwilliams) [inline dataflow/dags/comtrade_pipelines.py:82]: This one still allows nulls - should it be removed like the other or is there a different reason for this one?

**Comment by samuelhwilliams** [comment]
I look to be getting an error locally on the services pipeline. It scrapes 2019-2000 OK, but then tries to get 1999 and seems to be 404ing continuously (single example):

```
[2020-12-11 08:21:09,970] INFO {dataflow:comtrade.py:134} - Downloading https://comtrade.un.org/api/get/bulk/S/A/1999/all/EB02 []
[2020-12-11 08:21:26,777] INFO {backoff:_common.py:86} - Backing off _download(...) for 0.7s (requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://comtrade.un.org/api/get/bulk/S/A/1999/all/EB02?token=...)
```
